### PR TITLE
idp/openid: Automatically register OpenID-Connect users

### PR DESCRIPTION
**Pull request by mhilton**
Where ID tokens supply a preferred username automatically create the
user without requiring completion of the registration form where
possible. This also adds testing for the openid-connect IDP; and
prefills the full name and emails fields on the registration form where
possible.

### CODE REVIEW

File: idp/openid/openid-connect.go [hunk 1]
```diff
@@ -47,6 +47,25 @@ func init() {
 	})
 }
 
+// An IdentityCreator is used to create a candid identity from the
```
Comment (alesstimec) [inline idp/openid/openid-connect.go:50]: nitpick: s/An IdentityCreator/IdentityCreator/

File: idp/openid/openid-connect.go [hunk 2]
```diff
@@ -223,50 +248,72 @@ func (idp *openidConnectIdentityProvider) callback(ctx context.Context, w http.R
...
+			upd[store.Name] = store.Set
+		}
+		if user.Email != "" && existingUser.Email != user.Email {
+			existingUser.Email = user.Email
+			upd[store.Email] = store.Set
+		}
+		if (upd != store.Update{}) {
+			err = idp.initParams.Store.UpdateIdentity(ctx, &existingUser, upd)
+		}
+		if err == nil {
```
Comment (alesstimec) [inline idp/openid/openid-connect.go:279]: nitpick: this is a bit confusing as the only way to the err to be not nil comes from line 277..

**Review by alesstimec — APPROVED** [review]

File: idp/openid/openid-connect.go (hunk 1 above)
Comment (mhilton) [reply idp/openid/openid-connect.go:50]: The convention for documenting types is to start the comment with an indefinite article (e.g. https://golang.org/pkg/net/http/#Request)

File: idp/openid/openid-connect.go (hunk 2 above)
Comment (mhilton) [reply idp/openid/openid-connect.go:279]: yes but line #277 might not have been attempted, and if it was, and failed, we want to carry on with the function.

**Pull request merged**